### PR TITLE
OPAL: fix string buffer allocation for large env variables [4.0.x]

### LINKED ISSUE #8117: mpirun realloc() invalid next size
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

v4.0.0 and compiled with src code

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

from a source tarball

### Please describe the system on which you are running

* Operating system/version: ubuntu16.04
* Computer hardware: x86_64
* Network type: socket

-----------------------------

## Details of the problem

mpirun failed at `realloc`

```
*** Error in `mpirun': realloc(): invalid next size: 0x0000000001af4da0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7f59483ff7e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x834aa)[0x7f594840b4aa]
/lib/x86_64-linux-gnu/libc.so.6(realloc+0x179)[0x7f594840c839]
/usr/local/openmpi/lib/libopen-pal.so.40(opal_util_keyval_parse+0x459)[0x7f59489e0ac9]
/usr/local/openmpi/lib/libopen-pal.so.40(+0x5d08a)[0x7f59489cc08a]
/usr/local/openmpi/lib/libopen-pal.so.40(mca_base_var_cache_files+0x455)[0x7f59489cefe5]
/usr/local/openmpi/lib/libopen-pal.so.40(opal_init_util+0x1a2)[0x7f59489a7e32]
/usr/local/openmpi/lib/libopen-rte.so.40(orte_submit_init+0x19d)[0x7f5948d1774d]
mpirun[0x400fcf]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f59483a8830]
mpirun[0x400e39]
======= Memory map: ========
00400000-00402000 r-xp 00000000 fc:14 252069977                          /usr/local/openmpi/bin/orterun
00601000-00602000 r--p 00001000 fc:14 252069977                          /usr/local/openmpi/bin/orterun
00602000-00603000 rw-p 00002000 fc:14 252069977                          /usr/local/openmpi/bin/orterun
01ab7000-01afe000 rw-p 00000000 00:00 0                                  [heap]
7f5940000000-7f5940021000 rw-p 00000000 00:00 0 
7f5940021000-7f5944000000 ---p 00000000 00:00 0 
7f5947647000-7f5947658000 r-xp 00000000 fc:14 100813134                  /home/work/anaconda3/lib/libgcc_s.so.1
7f5947658000-7f5947857000 ---p 00011000 fc:14 100813134                  /home/work/anaconda3/lib/libgcc_s.so.1
7f5947857000-7f5947858000 r--p 00010000 fc:14 100813134                  /home/work/anaconda3/lib/libgcc_s.so.1
7f5947858000-7f5947859000 rw-p 00011000 fc:14 100813134                  /home/work/anaconda3/lib/libgcc_s.so.1
7f5947859000-7f594785b000 r-xp 00000000 fc:14 134324316                  /lib/x86_64-linux-gnu/libutil-2.23.so
7f594785b000-7f5947a5a000 ---p 00002000 fc:14 134324316                  /lib/x86_64-linux-gnu/libutil-2.23.so
7f5947a5a000-7f5947a5b000 r--p 00001000 fc:14 134324316                  /lib/x86_64-linux-gnu/libutil-2.23.so
7f5947a5b000-7f5947a5c000 rw-p 00002000 fc:14 134324316                  /lib/x86_64-linux-gnu/libutil-2.23.so
7f5947a5c000-7f5947b64000 r-xp 00000000 fc:14 134324280                  /lib/x86_64-linux-gnu/libm-2.23.so
7f5947b64000-7f5947d63000 ---p 00108000 fc:14 134324280                  /lib/x86_64-linux-gnu/libm-2.23.so
7f5947d63000-7f5947d64000 r--p 00107000 fc:14 134324280                  /lib/x86_64-linux-gnu/libm-2.23.so
7f5947d64000-7f5947d65000 rw-p 00108000 fc:14 134324280                  /lib/x86_64-linux-gnu/libm-2.23.so
7f5947d65000-7f5947d6c000 r-xp 00000000 fc:14 134324301                  /lib/x86_64-linux-gnu/librt-2.23.so
7f5947d6c000-7f5947f6b000 ---p 00007000 fc:14 134324301                  /lib/x86_64-linux-gnu/librt-2.23.so
7f5947f6b000-7f5947f6c000 r--p 00006000 fc:14 134324301                  /lib/x86_64-linux-gnu/librt-2.23.so
7f5947f6c000-7f5947f6d000 rw-p 00007000 fc:14 134324301                  /lib/x86_64-linux-gnu/librt-2.23.so
7f5947f6d000-7f5947f70000 r-xp 00000000 fc:14 134324261                  /lib/x86_64-linux-gnu/libdl-2.23.so
7f5947f70000-7f594816f000 ---p 00003000 fc:14 134324261                  /lib/x86_64-linux-gnu/libdl-2.23.so
7f594816f000-7f5948170000 r--p 00002000 fc:14 134324261                  /lib/x86_64-linux-gnu/libdl-2.23.so
7f5948170000-7f5948171000 rw-p 00003000 fc:14 134324261                  /lib/x86_64-linux-gnu/libdl-2.23.so
7f5948171000-7f5948187000 r-xp 00000000 fc:14 102174937                  /home/work/anaconda3/lib/libz.so.1.2.11
7f5948187000-7f5948386000 ---p 00016000 fc:14 102174937                  /home/work/anaconda3/lib/libz.so.1.2.11
7f5948386000-7f5948387000 r--p 00015000 fc:14 102174937                  /home/work/anaconda3/lib/libz.so.1.2.11
7f5948387000-7f5948388000 rw-p 00016000 fc:14 102174937                  /home/work/anaconda3/lib/libz.so.1.2.11
7f5948388000-7f5948548000 r-xp 00000000 fc:14 134324248                  /lib/x86_64-linux-gnu/libc-2.23.so
7f5948548000-7f5948748000 ---p 001c0000 fc:14 134324248                  /lib/x86_64-linux-gnu/libc-2.23.so
7f5948748000-7f594874c000 r--p 001c0000 fc:14 134324248                  /lib/x86_64-linux-gnu/libc-2.23.so
7f594874c000-7f594874e000 rw-p 001c4000 fc:14 134324248                  /lib/x86_64-linux-gnu/libc-2.23.so
7f594874e000-7f5948752000 rw-p 00000000 00:00 0 
7f5948752000-7f594876a000 r-xp 00000000 fc:14 134324297                  /lib/x86_64-linux-gnu/libpthread-2.23.so
7f594876a000-7f5948969000 ---p 00018000 fc:14 134324297                  /lib/x86_64-linux-gnu/libpthread-2.23.so
7f5948969000-7f594896a000 r--p 00017000 fc:14 134324297                  /lib/x86_64-linux-gnu/libpthread-2.23.so
7f594896a000-7f594896b000 rw-p 00018000 fc:14 134324297                  /lib/x86_64-linux-gnu/libpthread-2.23.so
7f594896b000-7f594896f000 rw-p 00000000 00:00 0 
7f594896f000-7f5948a76000 r-xp 00000000 fc:14 252071960                  /usr/local/openmpi/lib/libopen-pal.so.40.20.0
7f5948a76000-7f5948c76000 ---p 00107000 fc:14 252071960                  /usr/local/openmpi/lib/libopen-pal.so.40.20.0
7f5948c76000-7f5948c79000 r--p 00107000 fc:14 252071960                  /usr/local/openmpi/lib/libopen-pal.so.40.20.0
7f5948c79000-7f5948c7f000 rw-p 0010a000 fc:14 252071960                  /usr/local/openmpi/lib/libopen-pal.so.40.20.0
7f5948c7f000-7f5948c86000 rw-p 00000000 00:00 0 
7f5948c86000-7f5948d3a000 r-xp 00000000 fc:14 252071964                  /usr/local/openmpi/lib/libopen-rte.so.40.20.0
7f5948d3a000-7f5948f3a000 ---p 000b4000 fc:14 252071964                  /usr/local/openmpi/lib/libopen-rte.so.40.20.0
7f5948f3a000-7f5948f3c000 r--p 000b4000 fc:14 252071964                  /usr/local/openmpi/lib/libopen-rte.so.40.20.0
7f5948f3c000-7f5948f40000 rw-p 000b6000 fc:14 252071964                  /usr/local/openmpi/lib/libopen-rte.so.40.20.0
7f5948f40000-7f5948f43000 rw-p 00000000 00:00 0 
7f5948f43000-7f5948f44000 r-xp 00000000 fc:14 251660346                  /usr/lib/x86_64-linux-gnu/coreutils/libstdbuf.so
7f5948f44000-7f5949143000 ---p 00001000 fc:14 251660346                  /usr/lib/x86_64-linux-gnu/coreutils/libstdbuf.so
7f5949143000-7f5949144000 r--p 00000000 fc:14 251660346                  /usr/lib/x86_64-linux-gnu/coreutils/libstdbuf.so
7f5949144000-7f5949145000 rw-p 00001000 fc:14 251660346                  /usr/lib/x86_64-linux-gnu/coreutils/libstdbuf.so
7f5949145000-7f594916b000 r-xp 00000000 fc:14 134324228                  /lib/x86_64-linux-gnu/ld-2.23.so
7f5949354000-7f5949359000 rw-p 00000000 00:00 0 
7f5949367000-7f594936a000 rw-p 00000000 00:00 0 
7f594936a000-7f594936b000 r--p 00025000 fc:14 134324228                  /lib/x86_64-linux-gnu/ld-2.23.so
7f594936b000-7f594936c000 rw-p 00026000 fc:14 134324228                  /lib/x86_64-linux-gnu/ld-2.23.so
7f594936c000-7f594936d000 rw-p 00000000 00:00 0 
7ffde6726000-7ffde6748000 rw-p 00000000 00:00 0                          [stack]
7ffde679c000-7ffde679e000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]

+ POD_NAME=job754a8e08-worker-0
+ shift
+ /home/mine/kubectl exec job754a8e08-worker-0 -- /bin/sh -c '    PATH=/usr/local/openmpi/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/openmpi/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/openmpi/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/openmpi/bin/orted -mca ess "env" -mca ess_base_jobid "4133158912" -mca ess_base_vpid 1 -mca ess_base_num_procs "2" -mca orte_node_regex "job[3:754]a8e08-server-0,job[3:754]a8e08-worker-0@0(2)" -mca orte_hnp_uri "4133158912.0;tcp://172.16.0.63:47005" -mca pml "ob1" -mca btl "^openib" -mca plm_rsh_no_tree_spawn "true" -mca btl_tcp_if_include "192.168.0.0/16" -mca plm "rsh" -mca plm_rsh_agent "/home/mine/kubexec.sh" -mca orte_default_hostfile "/home/mine/hostfile" -mca pmix "^s1,s2,cray,isolated"'
--------------------------------------------------------------------------
ORTE has lost communication with a remote daemon.

  HNP daemon   : [[63067,0],0] on node job754a8e08-server-0
  Remote daemon: [[63067,0],1] on node job754a8e08-worker-0

This is usually due to either a failure of the TCP network
connection to the node, or possibly an internal failure of
the daemon itself. We cannot recover from this failure, and
therefore will terminate the job.
--------------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by devreal**
In OPAL's keyval parser in `add_to_env_str`, for the first variable a buffer of size 1024 was allocated, regardless of the space needed. For any additional variables, the size might have been doubled at best, even if more space was required. This PR makes sure the allocated buffer is sufficiently large to store the value.

Cherry-pick of https://github.com/open-mpi/ompi/pull/8144 to v4.0.x release branch.

Fixes #8117

Signed-off-by: Joseph Schuchart <schuchart@icl.utk.edu>

Also included cherry pick of #8163 (some coverity fixes that we found after we merged to master)

### CODE REVIEW

**Comment by jsquyres** [comment]
**NOTE:** This is not yet merged on master.  Only merge this PR after #8144 is merged.

**Comment by devreal** [comment]
Updated with latest change from https://github.com/open-mpi/ompi/pull/8144

**Review by jsquyres — APPROVED** [review]

**Comment by jsquyres** [comment]
Added commits from #8163 to this PR (i.e., cherry picked the Coverity fixes that were found after we merged to master).

**Pull request merged**
